pyro/optim/multi.py

```diff
@@ -0,0 +1,114 @@
+
+    Higher-order optimizers generally use :func:`torch.autograd.grad` rather
+    than :meth:`torch.Tensor.backward`, and therefore require a different
+    interface from usual Pyro and PyTorch optimizers. In this interface,
+    the :meth:`step` method inputs a ``loss`` tensor to be differentiated,
+    and backpropagation is triggered one or more times inside the optimizer.
+
+    Derived classes must implement :meth:`step` to compute derivatives and
+    update parameters in-place.
+
+    Example::
+
+        tr = poutine.trace(model).get_trace(*args, **kwargs)
+        loss = -tr.log_prob()
+        params = {name: pyro.param(name).unconstrained()
+                  for name in pyro.get_param_store().get_all_param_names()}
+        optim.step(loss, params)
+    """
+    def step(self, loss, params):
+        """
+        Performs an optimization step on parameters given a differentiable
+        ``loss`` tensor.
+
+        :param torch.Tensor loss: A differentiable tensor to be minimized.
+            Some optimizers require this to be differentiable multiple
+            times.
+        :param dict params: A dictionary mapping param name to unconstrained
+            value as stored in the param store.
+        """
+        raise NotImplementedError
+
+
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        names = []
+        values = []
+        for name, value in params.items():
+            names.append(name)
+            values.append(value)
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
+        self.optim(values)
+
+
+class TorchMultiOptimizer(PyroMultiOptimizer):
+    """
+    Facade to wrap :class:`~torch.optim.Optimizer` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim_constructor, optim_args):
+        optim = PyroOptim(optim_constructor, optim_args)
+        super(TorchMultiOptimizer, self).__init__(optim)
+
+
+class MixedMultiOptimizer(MultiOptimizer):
+    """
+    Container class to combine different :class:`MultiOptimizer` instances for
+    different parameters.
+
+    :param list parts: A list of ``(names, optim)`` pairs, where each
+        ``names`` is a list of parameter names, and each ``optim`` is a
+        :class:`MultiOptimizer` object to be used for the named parameters.
+        Together the ``names`` should partition up all desired parameters to
+        optimize.
+    """
+    def __init__(self, parts):
+        self.parts = list(parts)
+
+    def step(self, loss, params):
+        for names_part, optim in self.parts:
+            optim.step(loss, {name: params[name] for name in names_part})
```

Comment:
Does this behave correctly when each optimizer takes multiple derivatives of `loss`?  Does it require special care from the user or optimizer, e.g. using `retain_graph=True` in `grad` or not calling `backward`?

Reply:
Yes this should behave correctly since we've taken care to set `create_graph=True` when calling `grad()` in `PyroMultiOptimizer`. All derived classes of `MultiOptimizer` will set `create_graph=True` so they are composable.